v2/bundle/spiffebundle/bundle.go

```diff
@@ -3,175 +3,411 @@ package spiffebundle
 type Bundle struct {
+	trustDomain spiffeid.TrustDomain
+
+	mtx            sync.RWMutex
+	refreshHint    *int64
+	sequenceNumber *uint64
+	jwtKeys        map[string]crypto.PublicKey
+	x509Roots      []*x509.Certificate
 }
 
 // New creates a new bundle.
 func New(trustDomain spiffeid.TrustDomain) *Bundle {
-	panic("not implemented")
+	return &Bundle{
+		trustDomain: trustDomain,
+		jwtKeys:     make(map[string]crypto.PublicKey),
+	}
 }
 
 // Load loads a bundle from a file on disk.
 func Load(trustDomain spiffeid.TrustDomain, path string) (*Bundle, error) {
-	panic("not implemented")
+	bundleBytes, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, spiffebundleErr.New("unable to read SPIFFE bundle: %w", err)
+	}
+
+	return Parse(trustDomain, bundleBytes)
 }
 
 // Read decodes a bundle from a reader.
 func Read(trustDomain spiffeid.TrustDomain, r io.Reader) (*Bundle, error) {
-	panic("not implemented")
+	b, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, spiffebundleErr.New("unable to read: %v", err)
+	}
+
+	return Parse(trustDomain, b)
 }
 
 // Parse parses a bundle from bytes.
-func Parse(trustDomain spiffeid.TrustDomain, b []byte) (*Bundle, error) {
-	panic("not implemented")
+func Parse(trustDomain spiffeid.TrustDomain, bundleBytes []byte) (*Bundle, error) {
+	jwks := &bundleDoc{}
+	if err := json.Unmarshal(bundleBytes, jwks); err != nil {
+		return nil, spiffebundleErr.New("unable to parse JWKS: %v", err)
+	}
+
+	bundle := New(trustDomain)
+	if jwks.RefreshHint > 0 {
+		bundle.refreshHint = &jwks.RefreshHint
+	}
+	if jwks.SequenceNumber > 0 {
+		bundle.sequenceNumber = &jwks.SequenceNumber
+	}
+
+	if jwks.Keys == nil {
+		// The parameter keys MUST be present.
+		// https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#413-keys
+		return nil, spiffebundleErr.New("no keys found")
+	}
+	for i, key := range jwks.Keys {
+		switch key.Use {
+		// Two SVID types are supported: x509-svid and jwt-svid.
+		case x509SVIDUse:
+			if len(key.Certificates) != 1 {
+				return nil, spiffebundleErr.New("expected a single certificate in %s entry %d; got %d", x509SVIDUse, i, len(key.Certificates))
+			}
+			bundle.AddX509Root(key.Certificates[0])
+		case jwtSVIDUse:
+			if err := bundle.AddJWTKey(key.KeyID, key.Key); err != nil {
+				return nil, spiffebundleErr.New("error adding key %d of JWKS: %v", i, errs.Unwrap(err))
+			}
+		case "":
+			// The use parameter MUST be set.
+			// https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE_Trust_Domain_and_Bundle.md#422-public-key-use
+			return nil, spiffebundleErr.New("missing use for key entry %d", i)
+		}
+	}
+
+	return bundle, nil
 }
 
 // FromX509Bundle creates a bundle from an X.509 bundle.
 func FromX509Bundle(x509Bundle *x509bundle.Bundle) *Bundle {
-	panic("not implemented")
+	if x509Bundle != nil {
+		return &Bundle{
+			trustDomain: x509Bundle.TrustDomain(),
+			x509Roots:   x509Bundle.X509Roots(),
+		}
+	}
+	return &Bundle{}
```

Comment:
Unless there is a valid, ergonomic case for receiving a nil bundle, I'd advocate that we let it panic. Either way returning a valid Bundle pointer here seems like the wrong thing to do since it will not have a trust domain and will be unusable. If we decide not to panic, we should return `nil`.

Reply:
Makes sense, I'll change this to panic.